vulkan_generator/handle_remapper/generator.py

```diff
@@ -0,0 +1,323 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This is the entry point for Vulkan Code Generator"""
+
+import os
+
+from pathlib import Path
+from typing import List
+
+from textwrap import dedent
+from vulkan_parser import types
+from codegen_utils import codegen
+
+def is_dispatchable(handle: str) -> bool:
+    return len(handle) < 10 # Not actually the correct criteria for whether a handle is dispatchable
+
+def handle_map_name(handle: str) -> str:
+    return f"""{handle[2:3].lower() + handle[3:]}_handle_map_"""
+
+def handle_count_map_name(handle: str) -> str:
+    return f"""{handle[2:3].lower() + handle[3:]}_count_map_"""
+
+def handle_add_name(handle: str) -> str:
+    return f"""Add{handle[2:3].upper() + handle[3:]}Handle"""
+
+def handle_remove_name(handle: str) -> str:
+    return f"""Remove{handle[2:3].upper() + handle[3:]}Handle"""
+
+def handle_remap_name(handle: str) -> str:
+    return f"""Remap{handle[2:3].upper() + handle[3:]}Handle"""
+
+def dispatchable_handle_add_code(handle: str) -> str:
+    map_name = handle_map_name(handle)
+    return dedent(f"""
+        if({map_name}.find(captureHandle) != {map_name}.end()) throw HandleCollisionException();
+        {map_name}[captureHandle] = replayHandle;"""
+    )
+
+def dispatchable_handle_remove_code(handle: str) -> str:
+    map_name = handle_map_name(handle)
+    return dedent(f"""
+        if({map_name}.find(captureHandle) == {map_name}.end()) throw RemoveNonExistantHandleException();
+        {map_name}.erase(captureHandle);"""
+    )
+
+def dispatchable_handle_remap_code(handle: str) -> str:
+    map_name = handle_map_name(handle)
+    return dedent(f"""
+        if({map_name}.find(captureHandle) == {map_name}.end()) throw RemapNonExistantHandleException();
+        return {map_name}[captureHandle];"""
+    )
+
+def non_dispatchable_handle_add_code(handle: str) -> str:
+    map_name = handle_map_name(handle)
+    map_count_name = handle_count_map_name(handle)
+    return dedent(f"""
+        auto map_iter = {map_name}.find(captureHandle);
+        auto map_count_iter = {map_count_name}.find(captureHandle);
+
+        if(map_iter == {map_name}.end()) {{
+            if(map_count_iter != {map_count_name}.end()) throw InternalConsistencyException();
+        }}
+        else {{
+            if(map_count_iter == {map_count_name}.end() || map_count_iter->second <= 0) throw InternalConsistencyException();
+            if(map_iter->second != replayHandle) throw NonDispatchableHandleRedefinitionException();
+        }}
+
+        {map_name}[captureHandle] = replayHandle;
+        {map_count_name}[captureHandle] = {map_count_name}[captureHandle] +1;"""
+    )
+
+def non_dispatchable_handle_remove_code(handle: str) -> str:
+    map_name = handle_map_name(handle)
```

Comment:
These functions are great in terms of self explaining(as in you can just read the return string and see what this function generates and what's it's used for) but they still need some doc string to explain what they do

Reply:
Do we need doc strings like this?

""" generates the remove handle code for non dispatchable handles """
non_dispatchable_handle_remove_code() ...

Reply:
More like with an example of what is the input and what is the expected output. As its not clear to see what is the exact format we are outputting. 

Put yourself in the shoes of a person who don't know the interior of the code and wants to use this API. Then please think what kind of an explanation you would want.

Reply:
I've added this for the codegen functions.